File: go/internal/monitor/orchestrator.go
package monitor

import (
	"fmt"
	"sort"
	"strings"
	"sync"
)

func orchestrateCompany(company Company, prior *CompanyStatus) CrawlOutcome {
	plan := ResolveSourcePlan(company, prior)
	attempted := make([]string, 0, len(plan))
	var lastErr error
	var lastBlocked error

	for idx, source := range plan {
		attempted = append(attempted, source)
		jobs, err := FetchJobsForSource(company, source)
		if err == nil {
			qualityJobs := filterLikelyJobs(jobs)
			maxLinks := company.MaxLinks
			if maxLinks > 0 && len(qualityJobs) > maxLinks {
				qualityJobs = qualityJobs[:maxLinks]
			}
			hasMoreSources := idx < len(plan)-1
			if hasMoreSources && shouldTryFallbackForQuality(len(jobs), len(qualityJobs)) {
				lastErr = newFetchError("%s returned %d raw links but only %d looked like real jobs", source, len(jobs), len(qualityJobs))
				continue
			}
			relevantJobs := filterRelevantJobs(qualityJobs)
			message := fmt.Sprintf("%d job(s) fetched", len(relevantJobs))
			if len(relevantJobs) != len(jobs) {
				message = fmt.Sprintf(
					"%d job(s) fetched (%d raw links, %d quality-filtered, %d downstream-filtered)",
					len(relevantJobs),
					len(jobs),
					len(jobs)-len(qualityJobs),
					len(qualityJobs)-len(relevantJobs),
				)
			}
			return CrawlOutcome{
				Company:          company.Name,
				AttemptedSources: attempted,
				Status:           "ok",
				SelectedSource:   source,
				Jobs:             relevantJobs,
				Message:          message,
			}
		}
		if isBlockedErr(err) {
			lastBlocked = err
			continue
		}
		lastErr = err
	}

	if lastBlocked != nil && lastErr == nil {
		return CrawlOutcome{
			Company:          company.Name,
			AttemptedSources: attempted,
			Status:           "blocked",
			Message:          lastBlocked.Error(),
		}
	}
	if lastBlocked != nil && lastErr != nil {
		return CrawlOutcome{
			Company:          company.Name,
			AttemptedSources: attempted,
			Status:           "blocked",
			Message:          lastBlocked.Error() + " | secondary error: " + lastErr.Error(),
		}
	}
	message := "unknown error"
	if lastErr != nil {
		message = lastErr.Error()
	}
	return CrawlOutcome{
		Company:          company.Name,
		AttemptedSources: attempted,
		Status:           "error",
		Message:          message,
	}
}

func CrawlCompanies(
	companies []Company,
	workers int,
	prior map[string]CompanyStatus,
	onStart func(company string),
	onDone func(outcome CrawlOutcome),
) []CrawlOutcome {
	if len(companies) == 0 {
		return nil
	}
	if workers < 1 {
		workers = 1
	}
	if workers > len(companies) {
		workers = len(companies)
	}

	type workItem struct {
		Index   int
		Company Company
	}
	jobs := make(chan workItem)
	results := make(chan CrawlOutcome, len(companies))

	var wg sync.WaitGroup
	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for item := range jobs {
				if onStart != nil {
					onStart(item.Company.Name)
				}
				var priorStatus *CompanyStatus
				if st, ok := prior[item.Company.Name]; ok {
					copy := st
					priorStatus = &copy
				}
				_ = item.Index
				outcome := orchestrateCompany(item.Company, priorStatus)
				if onDone != nil {
					onDone(outcome)
				}
				results <- outcome
			}
		}()
	}

	for idx, company := range companies {
		jobs <- workItem{Index: idx, Company: company}
	}
	close(jobs)
	wg.Wait()
	close(results)

	outcomes := make([]CrawlOutcome, 0, len(companies))
	for result := range results {
		outcomes = append(outcomes, result)
	}

	sort.Slice(outcomes, func(i, j int) bool {
		return strings.ToLower(outcomes[i].Company) < strings.ToLower(outcomes[j].Company)
	})
	return outcomes
}

File: career_monitor/orchestrator.py
from __future__ import annotations

import logging
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any

from .adapters import fetch_jobs_for_source
from .exceptions import CrawlBlockedError
from .http_client import requests_session
from .models import CrawlOutcome
from .policy import base_source_plan, resolve_source_plan
from .utils import is_us_based_job, parse_bool_env


def orchestrate_company(
    company: dict[str, Any],
    prior_status: dict[str, Any] | None,
) -> CrawlOutcome:
    plan = resolve_source_plan(company, prior_status=prior_status)
    us_only_jobs = parse_bool_env("US_ONLY_JOBS", True)
    attempted: list[str] = []
    session = requests_session()
    last_error: Exception | None = None
    last_blocked: Exception | None = None

    for source in plan:
        attempted.append(source)
        try:
            jobs = fetch_jobs_for_source(company, source, session)
            if us_only_jobs:
                jobs = [
                    job
                    for job in jobs
                    if is_us_based_job(
                        title=job.title,
                        location=job.location,
                        url=job.url,
                    )
                ]
            return CrawlOutcome(
                company=company["name"],
                attempted_sources=attempted,
                status="ok",
                selected_source=source,
                jobs=jobs,
                message=f"{len(jobs)} job(s) fetched",
            )
        except CrawlBlockedError as exc:
            last_blocked = exc
            logging.warning("[%s] source '%s' blocked: %s", company["name"], source, exc)
            continue
        except Exception as exc:  # noqa: BLE001
            last_error = exc
            logging.warning("[%s] source '%s' failed: %s", company["name"], source, exc)
            continue

    if last_blocked is not None and last_error is None:
        return CrawlOutcome(
            company=company["name"],
            attempted_sources=attempted,
            status="blocked",
            message=str(last_blocked),
        )
    if last_blocked is not None and last_error is not None:
        return CrawlOutcome(
            company=company["name"],
            attempted_sources=attempted,
            status="blocked",
            message=f"{last_blocked} | secondary error: {last_error}",
        )
    return CrawlOutcome(
        company=company["name"],
        attempted_sources=attempted,
        status="error",
        message=str(last_error or "unknown error"),
    )


def crawl_companies(
    companies: list[dict[str, Any]],
    workers: int,
    prior_company_status: dict[str, dict[str, Any]],
) -> list[CrawlOutcome]:
    max_workers = max(1, min(workers, len(companies)))
    outcomes: list[CrawlOutcome] = []
    with ThreadPoolExecutor(max_workers=max_workers) as pool:
        future_to_company = {
            pool.submit(
                orchestrate_company,
                company,
                prior_company_status.get(company["name"]),
            ): company
            for company in companies
        }
        for future in as_completed(future_to_company):
            company = future_to_company[future]
            try:
                outcome = future.result()
            except Exception as exc:  # noqa: BLE001
                outcome = CrawlOutcome(
                    company=company["name"],
                    attempted_sources=base_source_plan(company),
                    status="error",
                    message=f"Unhandled worker exception: {exc}",
                )
            outcomes.append(outcome)
    outcomes.sort(key=lambda item: item.company.lower())
    return outcomes
